app/workers/tipline_newsletter_worker.rb

```diff
@@ -3,46 +3,71 @@ class TiplineNewsletterWorker
-          log team_id, language, 'Preparing newsletter to be sent...'
-          if !newsletter.nil? && Bot::Smooch.newsletter_content_changed?(newsletter, language, team_id)
-            date = I18n.l(Time.now.to_date, locale: language.to_s.tr('_', '-'), format: :long)
-            content = Bot::Smooch.build_newsletter_content(newsletter, language, team_id).gsub('{date}', date)
-            TiplineSubscription.where(language: language, team_id: team_id).find_each do |ts|
-              log team_id, language, "Sending newsletter to subscriber ##{ts.id}..."
-              begin
-                RequestStore.store[:smooch_bot_platform] = ts.platform
-                introduction = newsletter['smooch_newsletter_introduction'].to_s.gsub('{date}', date).gsub('{channel}', ts.platform)
-                content = content.gsub('{channel}', ts.platform)
-                Bot::Smooch.get_installation { |i| i.id == tbi.id }
-                response = Bot::Smooch.send_final_messages_to_user(ts.uid, content, workflow, language, 5)
-                Bot::Smooch.save_smooch_response(response, nil, nil, 'newsletter', language, { introduction: introduction })
-                log team_id, language, "Newsletter sent to subscriber ##{ts.id}, response: #{response.inspect}"
-                count += 1
-              rescue StandardError => e
-                log team_id, language, "Could not send newsletter to subscriber ##{ts.id}: #{e.message}"
-              end
-            end
-            tbi = TeamBotInstallation.find(tbi.id)
-            User.current = BotUser.smooch_user
-            newsletter = tbi.settings['smooch_workflows'].to_a.find{ |w| w['smooch_workflow_language'] == language }['smooch_newsletter']
-            tbi.skip_check_ability = true
-            newsletter['smooch_newsletter_last_sent_at'] = Time.now
-            tbi.save!
-            User.current = nil
-          else
-            log team_id, language, 'Not sending newsletter because content has not changed since the last delivery'
-          end
-        end
+    log team_id, language, 'Preparing newsletter to be sent...'
+    start = Time.now
+    TiplineSubscription.where(language: language, team_id: team_id).find_each do |ts|
+      log team_id, language, "Sending newsletter to subscriber ##{ts.id}..."
+      begin
+        RequestStore.store[:smooch_bot_platform] = ts.platform
+        Bot::Smooch.get_installation { |i| i.id == tbi.id }
+
+        response = Bot::Smooch.send_message_to_user(ts.uid, newsletter.format_as_template_message)
+
+        log team_id, language, "Newsletter sent to subscriber ##{ts.id}, response: (#{response.code}) #{response.body}"
+        count += 1
+      rescue StandardError => e
+        log team_id, language, "Could not send newsletter to subscriber ##{ts.id}: #{e.message}"
       end
     end
+    finish = Time.now
+
+    # Save a delivery event for this newsletter
+    TiplineNewsletterDelivery.create!({
```

Comment:
**question** this will raise an exception if creation fails for some reason, causing the job to get re-run from the beginning. is that intended and if so is this task idempotent?

Reply:
The task is configured to not be retried (some users might have received the newsletter already, and we don't want to resend)... I just changed to `create!` and to notify Sentry if this event can't be saved.